src/main/java/software/amazon/awssdk/crt/http/HttpConnection.java

```diff
@@ -0,0 +1,218 @@
+import software.amazon.awssdk.crt.io.ClientBootstrap;
+import software.amazon.awssdk.crt.io.SocketOptions;
+import software.amazon.awssdk.crt.io.TlsContext;
+
+import java.net.URI;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.util.concurrent.CompletableFuture;
+
+import static software.amazon.awssdk.crt.CRT.AWS_CRT_SUCCESS;
+
+
+/**
+ * This class wraps aws-c-http to provide the basic HTTP request/response functionality via the AWS Common Runtime.
+ *
+ * HttpConnection represents a single connection to a HTTP service endpoint.
+ *
+ * This class is not thread safe and should not be called from different threads.
+ */
+public class HttpConnection extends CrtResource {
+    private static final String HTTP = "http";
+    private static final String HTTPS = "https";
+
+    private final ClientBootstrap clientBootstrap;
+    private final SocketOptions socketOptions;
+    private final TlsContext tlsContext;
+    private final URI uri;
+    private final HttpConnectionEvents userConnectionCallbacks;
+
+    private ConnectionState connectionState = ConnectionState.DISCONNECTED;
+    private CompletableFuture<Boolean> connectedFuture;
+    private AsyncCallback connectAck;
+    private AsyncCallback disconnectAck;
+
+    public enum ConnectionState {
+        DISCONNECTED, CONNECTING, CONNECTED, SHUTTING_DOWN, SHUT_DOWN
+    }
+
+    /**
+     * Constructs a new HttpConnection. Connections are reusable after being disconnected.
+     * @param uri Must be non-null
+     * @throws HttpException If httpClient is null
+     */
+    public HttpConnection(URI uri) throws HttpException, CrtRuntimeException {
+        this(uri, new ClientBootstrap(1), new SocketOptions(), new TlsContext(), null);
+        own(this.clientBootstrap);
+        own(this.socketOptions);
+        own(this.tlsContext);
+    }
+
+    /**
+     * Constructs a new HttpConnection. Connections are reusable after being disconnected.
+     * @param uri Must be non-null
+     * @param callbacks Optional handler for connection interruptions/resumptions
+     * @throws HttpException If httpClient is null
+     */
+    public HttpConnection(URI uri, ClientBootstrap bootstrap, SocketOptions socketOptions, TlsContext tlsContext,
+                          HttpConnectionEvents callbacks) throws HttpException {
+
+        if (uri == null) {  throw new IllegalArgumentException("URI must not be null"); }
+        if (uri.getHost() == null) { throw new IllegalArgumentException("URI does not have a Host name"); }
+        if (bootstrap == null || bootstrap.isNull()) {  throw new IllegalArgumentException("ClientBootstrap must not be null"); }
+        if (socketOptions == null || socketOptions.isNull()) { throw new IllegalArgumentException("SocketOptions must not be null"); }
+        if ((HTTPS.equals(uri.getScheme()) || uri.getPort() == 443) && tlsContext == null) { throw new IllegalArgumentException("TlsContext must not be null if https is used"); }
```

Comment:
Is the `uri.getPort() == 443` part a safe assumption? What if some clients have a proxy so for some reason they connect over HTTP on port 443?

Reply:
That's a good catch, this will definitely fail when proxies get involved. We don't assume this anywhere other than unit tests at the moment, for that reason. If the user gives a bad port, that's on them, and the connection will fail accordingly.

Reply:
Updated Java side to require `http` or `https` Scheme for all URI's, and use the URI scheme to determine whether to use TLS or not. Java side then uses Scheme to decide whether or not to pass TlsContext/NULL to Native, and Native now always uses `TlsContext` if present.